cirq/sim/clifford_simulator.py

```diff
@@ -0,0 +1,749 @@
+            for i in range(2*self.n):
+                if i != p and self.xs[i,q]:
+                    self._rowsum(i,p)
+
+            self.xs[p-self.n,:] = self.xs[p,:]
+            self.zs[p-self.n,:] = self.zs[p,:]
+            self.rs[p-self.n] = self.rs[p]
+
+            self.xs[p,:] = False
+            self.zs[p,:] = False
+
+            self.zs[p,q] = True
+
+            self.rs[p] = bool(np.random.randint(2))
+
+            return int(self.rs[p])
+
+
+class CH_Form():
+    '''
+    A representation of stabilizer states using the CH form,
+
+        |psi> = omega U_C U_H |s>
+
+    This representation keeps track of overall phase.
+
+    See Bravyi et al, 2016 for details.
+    '''
+    def __init__(self,num_qubits,initial_state=0):
+        self.n = num_qubits
+        
+        # The state is represented by a set of binary matrices and vectors.
+        # See Section IVa of Bravyi et al
+        self.G = np.eye(self.n,dtype=bool)
+        
+        self.F = np.eye(self.n,dtype=bool)
+        self.M = np.zeros((self.n,self.n),dtype=bool)
+        self.gamma = np.zeros(self.n,dtype=int)
+        
+        self.v = np.zeros(self.n,dtype=bool)
+        self.s = np.zeros(self.n,dtype=bool)
+        
+        self.omega = 1
+
+        def bits(s):
+            while s > 0:
+                yield s & 1
+                s >>= 1
+
+        # Apply X for every non-zero element of initial_state
+        for (i,val) in enumerate(bits(initial_state)):
+            if val:
+                self._X(self.n-i-1)
+
+    def copy(self):
+        copy = CH_Form(self.n)
+
+        copy.G = self.G.copy()
+        copy.F = self.F.copy()
+        copy.M = self.M.copy()
+        copy.gamma = self.gamma.copy()
+        copy.v = self.v.copy()
+        copy.s = self.s.copy()
+        copy.omega = self.omega
+
+        return copy
+    
+    def __repr__(self):
+        ''' Return the wavefunction representation of the state. '''
+        return cirq.dirac_notation(self.to_numpy())  
+        
+    def _repr_full(self):
+        ''' Return the CH form representation of the state. '''
+        string = ""
+        
+        string += "omega: {:.2f}\n".format(self.omega)
+        string += "G:\n{}\n".format(np.array(self.G,dtype=int))
+        string += "F:\n{}\n".format(np.array(self.F,dtype=int))
+        string += "M:\n{}\n".format(np.array(self.M,dtype=int))
+        string += "gamma: {}\n".format(np.array(self.gamma,dtype=int))
+        string += "v: {}\n".format(np.array(self.v,dtype=int))
+        string += "s: {}\n".format(np.array(self.s,dtype=int))
+        
+        return string
+    
+    def _int2arr(self,x):
+        ''' Helper function that returns the bitstring representaiton of x '''
+        arr = np.zeros(self.n,dtype=bool)
+        
+        i = self.n-1
+        while x > 0:
+            if x & 1:
+                arr[i] = True
+            i -= 1
+            x >>= 1
+        
+        return arr
+        
+    def inner_product(self,x):
```

Comment:
This name is not very clear.

Add type annotations.

Reply:
Changed to `inner_product_of_state_and_x`.